core/src/main/java/io/questdb/cairo/vm/MemoryPARWImpl.java

```diff
@@ -465,193 +235,512 @@ public final void putLong128(long l1, long l2) {
+        }
+        return getLong0(offset);
+    }
+
+    /**
+     * Provides address of page for read operations. Memory writes never call this.
+     *
+     * @param page page index, starting from 0
+     * @return native address of page
+     */
+    public long getPageAddress(int page) {
+        if (page < pages.size()) {
+            return pages.getQuick(page);
+        }
+        return 0L;
+    }
+
+    public int getPageCount() {
+        return pages.size();
+    }
+
+    public long getPageSize() {
+        return getExtendSegmentSize();
+    }
+
+    public final short getShort(long offset) {
+        if (roOffsetLo < offset && offset < roOffsetHi - Short.BYTES) {
+            return Unsafe.getUnsafe().getShort(absolutePointer + offset);
+        }
+        return getShort0(offset);
+    }
+
+    public final CharSequence getStr(long offset) {
+        return getStr0(offset, csview);
+    }
+
+    public final CharSequence getStr2(long offset) {
+        return getStr0(offset, csview2);
+    }
+
+    public Long256 getLong256A(long offset) {
+        getLong256(offset, long256);
+        return long256;
+    }
+
+    public void getLong256(long offset, CharSink sink) {
+        final long a, b, c, d;
+        if (roOffsetLo < offset && offset < roOffsetHi - Long256.BYTES) {
+            a = Unsafe.getUnsafe().getLong(absolutePointer + offset);
+            b = Unsafe.getUnsafe().getLong(absolutePointer + offset + Long.BYTES);
+            c = Unsafe.getUnsafe().getLong(absolutePointer + offset + Long.BYTES * 2);
+            d = Unsafe.getUnsafe().getLong(absolutePointer + offset + Long.BYTES * 3);
         } else {
-            putLong(offset, l0);
-            putLong(offset + Long.BYTES, l1);
-            putLong(offset + Long.BYTES * 2, l2);
-            putLong(offset + Long.BYTES * 3, l3);
+            a = getLong(offset);
+            b = getLong(offset + Long.BYTES);
+            c = getLong(offset + Long.BYTES * 2);
+            d = getLong(offset + Long.BYTES * 3);
         }
+        Numbers.appendLong256(a, b, c, d, sink);
     }
 
-    @Override
-    public final void putLong256(long l0, long l1, long l2, long l3) {
-        if (pageHi - appendPointer > Long256.BYTES - 1) {
-            Unsafe.getUnsafe().putLong(appendPointer, l0);
-            Unsafe.getUnsafe().putLong(appendPointer + Long.BYTES, l1);
-            Unsafe.getUnsafe().putLong(appendPointer + Long.BYTES * 2, l2);
-            Unsafe.getUnsafe().putLong(appendPointer + Long.BYTES * 3, l3);
-            appendPointer += Long256.BYTES;
+    public void getLong256(long offset, Long256Acceptor sink) {
+        if (roOffsetLo < offset && offset < roOffsetHi - Long256.BYTES) {
+            sink.setAll(
+                    Unsafe.getUnsafe().getLong(absolutePointer + offset),
+                    Unsafe.getUnsafe().getLong(absolutePointer + offset + Long.BYTES),
+                    Unsafe.getUnsafe().getLong(absolutePointer + offset + Long.BYTES * 2),
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).